### ISSUE TITLE
Can't figure out how to use a recursive bounded generic type

### ISSUE THREAD

**Issue by myronmarston** [soutaro/steep#603]
Given `Steepfile`:

``` ruby
target :lib do
  signature "sig"
  check "lib"                       # Directory name
  library "set"
end
```

And `sig/set_value_extractor.rbs`:

``` rbs
# Define the minimal operations needed by any `Set` implementation. It takes a type parameter `S`
# because the operators each take an arg of the same type and return that type. e.g. a given set
# implementation only has to support working with arguments of the same type (and returns results
# of that same type).
interface _MinimalSet[S]
  def intersection: (S) -> S
  def union: (S) -> S
  def -: (S) -> S
end

# Here we define a class that can work with multiple set implementations, bounded by the
# _MinimalSet interface. We use `S` recursively: it's both the type parameter of `SetValueExtractor`
# and is also the type parameter of `_MinimalSet`.
class SetValueExtractor[S < _MinimalSet[S]]
  def build_set: (:union | :intersection | :-, S, S) -> S
end

class SomeAction
  # Here we try to define an instance variable of `SetValueExtractor` that operators on integer
  # sets, but Steep doesn't like this :(.
  @set_value_extractor: SetValueExtractor[::Set[::Integer]]
end
```

And `lib/set_value_extractor.rb`:

``` ruby
class SetValueExtractor
  def build_set(operator, set1, set2)
    case operator
    when :union
      set1.union(set2)
    when :intersection
      set1.intersection(set2)
    when :-
      set1 - set2
    end
  end
end

class SomeAction
  def initialize
    @set_value_extractor = SetValueExtractor.new
  end
end
```

....when I run `steep check`, I get:

```
sig/set_value_extractor.rbs:12:24: [error] Type application of `::SetValueExtractor` doesn't satisfy the constraints: ::Set[::Integer] <: ::_MinimalSet[S]
│ Diagnostic ID: RBS::UnsatisfiableTypeApplication
│
└   @set_value_extractor: SetValueExtractor[::Set[::Integer]]
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Is this intended to work?  If not, how do you recommend bounding a generic type with an interface like `_MinimalSet` that only accepts args and returns values of the same type as the implementation type?

**Comment by soutaro** [comment]
Hi @myronmarston,

Thank you for reporting this problem. Confirmed this is a bug. 🙇‍♂️ 

RBS allows recursive constraints on generics, Steep should apply the type application before validating the correctness.
I've written a patch to fix the problem, and will consider if it's correct way to handle this.

```diff
diff --git a/lib/steep/signature/validator.rb b/lib/steep/signature/validator.rb
index 0ee2f2b..646ba00 100644
--- a/lib/steep/signature/validator.rb
+++ b/lib/steep/signature/validator.rb
@@ -58,9 +58,14 @@ module Steep
 
       def validate_type_application_constraints(type_name, type_params, type_args, location:)
         if type_params.size == type_args.size
+          subst = Interface::Substitution.build(
+            type_params.map(&:name),
+            type_args.map {|type| factory.type(type) }
+          )
+
           type_params.zip(type_args).each do |param, arg|
             if param.upper_bound
-              upper_bound_type = factory.type(param.upper_bound)
+              upper_bound_type = factory.type(param.upper_bound).subst(subst)
               arg_type = factory.type(arg)
 
               constraints = Subtyping::Constraints.empty
```

**Comment by soutaro** [comment]
Released [1.0.2](https://rubygems.org/gems/steep/versions/1.0.2) with this fix. 
Will be back ported in 1.1 too.
